src/apps/vpn/productshandler.cpp

```diff
@@ -41,6 +45,36 @@ ProductsHandler::ProductsHandler(QObject* parent) : QAbstractListModel(parent) {
   MZ_COUNT_CTOR(ProductsHandler);
   Q_ASSERT(!s_instance);
   s_instance = this;
+
+  // On iAP make sure the Products are loaded in time.
+  // If we Move into any State adjecent to iAP - load the
+  // products if we don't have that already.
+  connect(MozillaVPN::instance(), &MozillaVPN::stateChanged, this, [this]() {
+    auto state = MozillaVPN::instance()->state();
+    if ((state == App::StateSubscriptionNeeded ||
+         state == App::StateSubscriptionInProgress ||
+         state == App::StateAuthenticating) &&
+        !this->hasProductsRegistered()) {
+      TaskScheduler::scheduleTask(new TaskProducts());
+    }
+    if (state == App::StateMain) {
```

Comment:
Totally agree with the previous `if` block. More than agree. It's clever and I really like it.

But, I don't see why the below `if` block is here. It doesn't seem like an obvious place to keep this....meaning I wouldn't know to look here for this logic. What about putting this at the end of `maybeStateMain`?

Reply:
I was asked to move that here :) - https://github.com/mozilla-mobile/mozilla-vpn-client/pull/6488#discussion_r1152693162

I think the idea as a first step to move things out of the mozillvpn.cpp that could be reused, for you know that _thing_.
Given that is not happening - how about we move that piece back into it's original place in `void MozillaVPN::accountChecked` - Given that can only happen due to an account check?